### PR TITLE
[LV] Refactor vector function variant selection to prepare for uniform args

### PR DESCRIPTION
**Pull request by huntergr-arm**
Parameters marked as uniform take a scalar value, assuming the value is
invariant in the scalar loop.

### CODE REVIEW

**Comment by huntergr-arm** [comment]
I have a followup patch to support linear arguments once this has been merged.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 1]
```diff
@@ -7009,39 +7009,60 @@ void LoopVectorizationCostModel::setVectorizedCallDecision(ElementCount VF) {
...
+          case VFParamKind::Vector:
+            break;
+          case VFParamKind::OMP_Uniform: {
+            Value *ScalarParam = CI->getArgOperand(Param.ParamPos);
+            // Make sure the scalar parameter in the loop is invariant.
+            if (!PSE.getSE()->isLoopInvariant(PSE.getSCEV(ScalarParam),
+                                              TheLoop))
+              ParamsOk = false;
+            break;
+          }
```
Comment (mgabka) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:7038]: to me looks like the piece of code handling "case VFParamKind::OMP_Uniform: " is the actual new functionality, so you could split this patch into NFC change/refactoring + new functionality+tests demonstrating that it works, what do you think?

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 2]
```diff
@@ -7009,39 +7009,60 @@ void LoopVectorizationCostModel::setVectorizedCallDecision(ElementCount VF) {
...
-        assert(MaskPos.has_value() && "Unable to find mask parameter index");
+        // Found a suitable candidate, stop here.
+        VecFunc = CI->getModule()->getFunction(Info.VectorName);
+        FuncInfo = Info;
+        break;
       }
 
+      // Add in the cost of synthesizing a mask if one wasn't required.
+      InstructionCost MaskCost = 0;
+      if (VecFunc && UsesMask && !MaskRequired)
```
Comment (mgabka) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:7059]: this isn't functionality related to support of the uniform paramaters, and in my opinion should go as a separate patch with extra tests checking the increased cost.

File: llvm/test/Transforms/LoopVectorize/AArch64/uniform-args-call-variants.ll [hunk 3]
```diff
@@ -0,0 +1,117 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 2
+; RUN: opt < %s -passes=loop-vectorize,instcombine -force-vector-interleave=1 -S | FileCheck %s
+
+target triple = "aarch64-unknown-linux-gnu"
+
+; A call whose argument can remain a scalar for a vectorized function variant
+; with a uniform argument because it's loop invariant
+define void @test_uniform(ptr noalias %dst, ptr readonly %src, i64 %uniform , i64 %n) #0 {
```
[the remark this answers is not in the document]
Comment (huntergr-arm) [reply llvm/test/Transforms/LoopVectorize/AArch64/uniform-args-call-variants.ll:8]: https://github.com/llvm/llvm-project/pull/72260 adds support for mixing element types.

I'll add some masking tests.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 1 above)
Comment (huntergr-arm) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:7038]: Yeah, the reason for putting it all together was the lack of dependent commits for review chains. That's now been resolved, though I don't think I can auto convert this PR to a branch on the main repo. Will raise a new PR for the uniform bit.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 2 above)
Comment (huntergr-arm) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:7059]: That's just refactoring of existing code, not new functionality.

**Comment by huntergr-arm** [comment]
I've removed the uniform argument case and the tests from this PR, so it should just be NFC refactoring in preparation now. I'll precommit the tests then raise a new PR for the uniform case. Ok?

**Comment by mgabka** [comment]
> I've removed the uniform argument case and the tests from this PR, so it should just be NFC refactoring in preparation now. I'll precommit the tests then raise a new PR for the uniform case. Ok?

Thanks for extracting the NFC refactoring, it LGTM

**Review by mgabka — APPROVED** [review]
